SU2_CFD/src/numerics_direct_turbulent.cpp

```diff
@@ -479,6 +479,480 @@ void CSourcePieceWise_TurbSA::ComputeResidual(su2double *val_residual, su2double
+    cb2_sigma = cb2/sigma;
+    cw1 = cb1/k2+(1.0+cb2)/sigma;
+    
+}
+
+CSourcePieceWise_TurbSA_E::~CSourcePieceWise_TurbSA_E(void) { }
+
+void CSourcePieceWise_TurbSA_E::ComputeResidual(su2double *val_residual, su2double **val_Jacobian_i, su2double **val_Jacobian_j, CConfig *config) {
+    
+    //  AD::StartPreacc();
+    //  AD::SetPreaccIn(V_i, nDim+6);
+    //  AD::SetPreaccIn(Vorticity_i, nDim);
+    //  AD::SetPreaccIn(StrainMag_i);
+    //  AD::SetPreaccIn(TurbVar_i[0]);
+    //  AD::SetPreaccIn(TurbVar_Grad_i[0], nDim);
+    //  AD::SetPreaccIn(Volume); AD::SetPreaccIn(dist_i);
+    
+    if (incompressible) {
+        Density_i = V_i[nDim+1];
+        Laminar_Viscosity_i = V_i[nDim+3];
+    }
+    else {
+        Density_i = V_i[nDim+2];
+        Laminar_Viscosity_i = V_i[nDim+5];
+    }
+    
+    val_residual[0] = 0.0;
+    Production      = 0.0;
+    Destruction     = 0.0;
+    CrossProduction = 0.0;
+    val_Jacobian_i[0][0] = 0.0;
+    
+    
+    /*
+     From NASA Turbulence model site. http://turbmodels.larc.nasa.gov/spalart.html
+     This form was developed primarily to improve the near-wall numerical behavior of the model (i.e., the goal was to improve the convergence behavior). The reference is:
+     Edwards, J. R. and Chandra, S. "Comparison of Eddy Viscosity-Transport Turbulence Models for Three-Dimensional, Shock-Separated Flowfields," AIAA Journal, Vol. 34, No. 4, 1996, pp. 756-763.
+     In this modificaton Omega is replaced by Strain Rate
+     */
+    
+    /*--- Evaluate Omega, here Omega is the Strain Rate ---*/
+    
+    Sbar = 0.0;
+    for(iDim=0;iDim<nDim;++iDim){
+        for(jDim=0;jDim<nDim;++jDim){
+            Sbar+= (PrimVar_Grad_i[1+iDim][jDim]+PrimVar_Grad_i[1+jDim][iDim])*(PrimVar_Grad_i[1+iDim][jDim]);}}
+    for(iDim=0;iDim<nDim;++iDim){
+        Sbar-= ((2.0/3.0)*pow(PrimVar_Grad_i[1+iDim][iDim],2.0));}
+    
+    Omega= sqrt(max(Sbar,0.0));
+    
+    /*--- Evaluate Omega ---*/
+    
+    //Omega = sqrt(Vorticity_i[0]*Vorticity_i[0] + Vorticity_i[1]*Vorticity_i[1] + Vorticity_i[2]*Vorticity_i[2]);
+    
+    /*--- Rotational correction term ---*/
+    
+    if (rotating_frame) { Omega += 2.0*min(0.0, StrainMag_i-Omega); }
+    
+    if (dist_i > 1e-10) {
+        
+        /*--- Production term ---*/
+        
+        dist_i_2 = dist_i*dist_i;
+        nu = Laminar_Viscosity_i/Density_i;
+        Ji = TurbVar_i[0]/nu;
+        Ji_2 = Ji*Ji;
+        Ji_3 = Ji_2*Ji;
+        fv1 = Ji_3/(Ji_3+cv1_3);
+        fv2 = 1.0 - Ji/(1.0+Ji*fv1);
+        ft2 = ct3*exp(-ct4*Ji_2);
+        S = Omega;
+        inv_k2_d2 = 1.0/(k2*dist_i_2);
+        
+        //Shat = S + TurbVar_i[0]*fv2*inv_k2_d2;
+        Shat = max(S*((1.0/max(Ji,1.0e-16))+fv1),1.0e-16);
+        
+        Shat = max(Shat, 1.0e-10);
+        inv_Shat = 1.0/Shat;
+        
+        /*--- Production term ---*/;
+        
+        //    Original SA model
+        //    Production = cb1*(1.0-ft2)*Shat*TurbVar_i[0]*Volume;
```

Comment:
In general, please remove the commented lines from the original S-A model to keep the implementation as clean as possible (especially for new users).